stellargraph/connector/neo4j/mapper.py

```diff
@@ -266,27 +202,21 @@ def sample_features(self, head_nodes, batch_num):
             given the sequence of in/out directions.
         """
         node_samples = self.sampler.run(
-            self.neo4j_graphdb,
-            nodes=head_nodes,
-            n=1,
-            in_size=self.in_samples,
-            out_size=self.out_samples,
+            nodes=head_nodes, n=1, in_size=self.in_samples, out_size=self.out_samples,
         )
 
         # Reshape node samples to sensible format
         # Each 'slot' represents the list of nodes sampled from some neighbourhood, and will have a corresponding
         # NN input layer. Every hop potentially generates both in-nodes and out-nodes, held separately,
         # and thus the slot (or directed hop sequence) structure forms a binary tree.
 
-        node_type = self.head_node_types[0]
-
         max_hops = len(self.in_samples)
         max_slots = 2 ** (max_hops + 1) - 1
         features = [None] * max_slots  # flattened binary tree
 
         for slot in range(max_slots):
             nodes_in_slot = node_samples[slot]
-            features_for_slot = self.graph.node_features(nodes_in_slot, node_type)
+            features_for_slot = self.graph.node_features(nodes_in_slot)
```

Comment:
Feature lookups seem to be the most expensive op here, it might be worth doing something like:

```python
batch_nodes = np.concatenate(node_samples)
batch_features = self.graph.node_features(batch_nodes)
idx = 0
for slot in range(max_slots):
    features_for_slot = batch_features[idx: (idx+len(node_samples[slot])]
    idx += len(node_samples[slot])
```

Reply:
^ hopefully that reduces some of the query overhead but i'm not sure whether it'll be noticeable 🤷‍♂️

Reply:
Thanks Kieran! I've updated it to use your suggestion, the benefits are probably gonna be more noticeable when the node features aren't cached though